pkg/workflow/controllers/chaos_node_reconciler.go

```diff
@@ -0,0 +1,206 @@
+	ctx := context.TODO()
+	node := v1alpha1.WorkflowNode{}
+
+	err := it.kubeClient.Get(ctx, request.NamespacedName, &node)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if ConditionEqualsTo(node.Status, v1alpha1.ConditionDeadlineExceed, corev1.ConditionTrue) {
+		err := it.recoverChaos(ctx, node)
+		return reconcile.Result{}, err
+	}
+
+	if availableChaos(string(node.Spec.Type)) {
+		err = it.injectChaos(ctx, node)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (it *ChaosNodeReconciler) injectChaos(ctx context.Context, node v1alpha1.WorkflowNode) error {
+	var chaosObject runtime.Object
+	var meta metav1.Object
+
+	if node.Spec.Type == v1alpha1.TypeNetworkChaos {
+		networkChaos := v1alpha1.NetworkChaos{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", node.Name),
+				Namespace:    node.Namespace,
+			},
+			Spec: *node.Spec.NetworkChaos,
+		}
+		meta = networkChaos.GetObjectMeta()
+		chaosObject = &networkChaos
+	} else if node.Spec.Type == v1alpha1.TypePodChaos {
+		podChaos := v1alpha1.PodChaos{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", node.Name),
+				Namespace:    node.Namespace,
+			},
+			Spec: *node.Spec.PodChaos,
+		}
+		meta = podChaos.GetObjectMeta()
+		chaosObject = &podChaos
+	} else {
+		// TODO: support other type of chaos by code generation
+		it.logger.Info("unsupported chaos nodes", "key", fmt.Sprintf("%s/%s", node.Namespace, node.Name), "type", node.Spec.Type)
+		return nil
+	}
+
+	meta.SetOwnerReferences(append(meta.GetOwnerReferences(), metav1.OwnerReference{
+		APIVersion:         node.APIVersion,
+		Kind:               node.Kind,
+		Name:               node.Name,
+		UID:                node.UID,
+		Controller:         &isController,
+		BlockOwnerDeletion: &blockOwnerDeletion,
+	}))
+
+	err := it.kubeClient.Create(ctx, chaosObject)
+	if err != nil {
+		it.eventRecorder.Event(&node, corev1.EventTypeWarning, v1alpha1.ChaosCRCreateFailed, "Failed to create chaos CR")
+		it.logger.Error(err, "failed to create chaos")
+		return nil
+	}
+
+	it.eventRecorder.Event(&node, corev1.EventTypeNormal, v1alpha1.ChaosCRCreated, fmt.Sprintf("Chaos CR %s/%s created", meta.GetNamespace(), meta.GetName()))
+
+	group := chaosObject.GetObjectKind().GroupVersionKind().Group
+	chaosRef := corev1.TypedLocalObjectReference{
+		APIGroup: &group,
+		Kind:     chaosObject.GetObjectKind().GroupVersionKind().Kind,
+		Name:     meta.GetName(),
+	}
+
+	updateError := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		node := v1alpha1.WorkflowNode{}
+		err := it.kubeClient.Get(ctx, types.NamespacedName{
+			Namespace: node.Namespace,
+			Name:      node.Name,
+		}, &node)
+		if err != nil {
+			return client.IgnoreNotFound(err)
+		}
+		node.Status.ChaosResource = &chaosRef
+
+		// TODO: this condition should be set by observation
```

Comment:
what is the `observation` means?

Reply:
It's related to the above comments about labeling resources when creating.

Currently, we set that `ConditionChaosInjected` to `true`, because we have created a Chaos Object. It will cause the similar issue with https://github.com/chaos-mesh/chaos-mesh/pull/1551#discussion_r597441206. 

The best practice is pretty same with the above comment: labeling when creating resource, and if we could list out some resource with that label, change that condition to true.